agent/replacecall/src/main/java/randoop/instrument/CallReplacementTransformer.java

```diff
@@ -0,0 +1,398 @@
+          ReplaceCallAgent.debugPath + File.separator + "transform-log.txt",
+          ReplaceCallAgent.debug);
+
+  private static SimpleLog debug_instrument_inst =
+      new SimpleLog(
+          ReplaceCallAgent.debugPath + File.separator + "instrument-log.txt",
+          ReplaceCallAgent.debug);
+
+  /** Debug information on method maping */
+  private static SimpleLog debug_map =
+      new SimpleLog(
+          ReplaceCallAgent.debugPath + File.separator + "method_mapping.txt",
+          ReplaceCallAgent.debug);
+
+  /** Map from a method to its replacement. */
+  private final ConcurrentHashMap<MethodSignature, MethodSignature> replacementMap;
+
+  /** The list of package prefixes (package name + ".") to exclude from transformation. */
+  private final Set<String> excludedPackagePrefixes;
+
+  /**
+   * Create a {@link CallReplacementTransformer} that transforms method calls in classes other than
+   * those named in the given exclusion set.
+   *
+   * <p>The transformer can be run by multiple threads, so the replacement maps use concurrent
+   * implementations.
+   *
+   * @param replacementMap the concurrent hash map with method replacements
+   * @param excludedPackagePrefixes the period-terminated prefixes for packages from which classes
+   *     should not be transformed
+   */
+  CallReplacementTransformer(
+      ConcurrentHashMap<MethodSignature, MethodSignature> replacementMap,
+      Set<String> excludedPackagePrefixes) {
+    this.replacementMap = replacementMap;
+    this.excludedPackagePrefixes = excludedPackagePrefixes;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Transforms the given class class by replacing calls to methods with corresponding calls as
+   * determined by {@link #replacementMap}.
+   *
+   * <p>Excludes bootloaded classes that are not AWT/Swing classes. Other exclusions are determined
+   * by the set of {@link #excludedPackagePrefixes}.
+   *
+   * @see ReplaceCallAgent
+   */
+  @Override
+  public byte[] transform(
+      ClassLoader loader,
+      String className,
+      Class<?> classBeingRedefined,
+      ProtectionDomain protectionDomain,
+      byte[] classfileBuffer)
+      throws IllegalClassFormatException {
+    // Note: An uncaught exception within a transform method results in null being returned.
+
+    String fullClassName = className.replace("/", ".");
+
+    if (isExcludedClass(loader, fullClassName)) {
+      debug_transform.log("transform: ignoring excluded class %s%n", className);
+      return null;
+    }
+
+    debug_transform.log("%ntransform: ENTER %s%n", className);
+
+    // Parse the bytes of the classfile
+    JavaClass c;
+    try {
+      ClassParser parser = new ClassParser(new ByteArrayInputStream(classfileBuffer), className);
+      c = parser.parse();
+    } catch (Exception e) {
+      debug_transform.log("transform: EXIT parse of %s resulted in error %s%n", className, e);
+      throw new RuntimeException("Unexpected parse exception", e);
```

Comment:
Because of how exceptions in `ClassFileTransform.transform()` are handled there is no useful information that comes from throwing the exception.